File: Semestr 3/Paradygmaty_Programowania/Scala/src/lab_2.scala
// zadanie 2
def nTy[A] (list:List[A], index:Int):A =
  index match{
  case x if x < 1 || list == Nil => throw new Exception("Niepoprawny index")
  case 1 => list.head
  case x => nTy(list.tail, index-1)
}

// zadanie 4
def podzielPoN[A] (lista:List[A], dlugosc:Int):(List[A], List[A]) =
  (lista, dlugosc) match{
  case (Nil, x) if x > 0 => throw new Exception("Niepoprawna dlugosc")
  case (_, x) if x < 0 => throw new Exception("Niepoprawna dlugosc")
  case (l, x) if x == 0 => (Nil, l)
  case (h::t, x) => {val (a,b) = podzielPoN(t, x-1); (h::a, b) }
}










def pocz[A] (lista:List[A], dlugosc:Int):List[A] =
  (lista, dlugosc) match{
  case (_, x) if x <= 0 => Nil
  case (Nil, x) if x > 0 => Nil
  case (h::t, x) => h::pocz(t,x-1)
}

def kon[A] (lista:List[A], dlugosc:Int):List[A] =
  (lista, dlugosc) match{
  case (l, x) if x <= 0 => l
  case (Nil, x) if x > 0 =>Nil
  case (h::t, x) => kon(t, x-1)
}

def podzielPoN[A] (lista:List[A], dlugosc:Int):(List[A], List[A]) = {
  (pocz(lista, dlugosc), kon(lista,dlugosc))
}

File: Semestr 3/Paradygmaty_Programowania/Ocaml/lab_2.ml
(* zadanie 1 *)
let rec nTy (list, index) =
	match index with
	| x when x < 1 || list = [] -> failwith "Niepoprawny index" 
	| 1 -> List.hd list
	| x -> nTy(List.tl list, index -1)
	
	
	
	
	
	
(* zadanie 3*)
	let rec podzielPoN (lista, dlugosc) =
		match (lista, dlugosc) with
		| ([],x) when x > 0 -> failwith "Niepoprawna dlugosc" 
		| (_, x) when x < 0 -> failwith "Niepoprawna dlugosc"
		| (l, x) when x = 0 -> ([],l)
		| (h::t,x) -> let (a,b) = podzielPoN(t, x-1) in (h::a, b)



let rec podzielPoN (lista, dlugosc) =
	let rec pocz(l,x) = 
		match (l,x) with
		| (_, y) when y <= 0 -> []
		| ([], y) when y >0 -> []
		| (h::t, y) -> h::pocz(t,y-1)
	in
	let rec kon(list,x) =
		match (list,x) with
		| (l, x) when x <= 0 -> l
		| ([], x) when x > 0 -> []
		| (h::t, y) -> kon(t, y-1) 
	in
	(pocz(lista, dlugosc), kon(lista,dlugosc))

podzielPoN([1;2;3;4],(-1));;
podzielPoN([1;2;3;4],0);;
podzielPoN([1;2;3;4],1);;
podzielPoN([1;2;3;4],2);;
podzielPoN([1;2;3;4],3);;
podzielPoN([1;2;3;4],4);;
podzielPoN([1;2;3;4],5);;

		
		
		
		